In [1]:
import numpy as np
import pandas as pd

### 1. Load data và xuất 5 bản ghi đầu tiên của data

In [2]:
df = pd.read_csv('countries of the world.csv')
df.head()

Country                               Region  Population  \
0     Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1         Albania   EASTERN EUROPE                          3581655   
2         Algeria   NORTHERN AFRICA                        32930091   
3  American Samoa   OCEANIA                                   57794   
4         Andorra   WESTERN EUROPE                            71201   

   Area (sq. mi.) Pop. Density (per sq. mi.) Coastline (coast/area ratio)  \
0          647500                       48,0                         0,00   
1           28748                      124,6                         1,26   
2         2381740                       13,8                         0,04   
3             199                      290,4                        58,29   
4             468                      152,1                         0,00   

  Net migration Infant mortality (per 1000 births)  GDP ($ per capita)  \
0         23,06                             163,07               700.0   
1         -4,93                              21,52              4500.0   
2         -0,39                                 31              6000.0   
3        -20,71                               9,27              8000.0   
4           6,6                               4,05             19000.0   

  Literacy (%) Phones (per 1000) Arable (%) Crops (%) Other (%) Climate  \
0         36,0               3,2      12,13      0,22     87,65       1   
1         86,5              71,2      21,09      4,42     74,49       3   
2         70,0              78,1       3,22      0,25     96,53       1   
3         97,0             259,5         10        15        75       2   
4        100,0             497,2       2,22         0     97,78       3   

  Birthrate Deathrate Agriculture Industry Service  
0      46,6     20,34        0,38     0,24    0,38  
1     15,11      5,22       0,232    0,188   0,579  
2     17,14      4,61       0,101      0,6   0,298  
3     22,46      3,27         NaN      NaN     NaN  
4      8,71      6,25         NaN      NaN     NaN

### 2. Lấy dữ liệu từ các cột
Lấy dữ liệu từ các cột sau **'Country', 'Region', 'Population', 'Area (sq. mi.)', 'Coastline (coast/area ratio)', 'GDP ( per capita)'** (các yêu cầu dưới sẽ thực hiện trên dữ liệu này).

In [3]:
cols = ['Country', 'Region', 'Population', 'Area (sq. mi.)', 'Coastline (coast/area ratio)', 'GDP ($ per capita)']
countries_df = df[cols]
countries_df.head()

Country                               Region  Population  \
0     Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1         Albania   EASTERN EUROPE                          3581655   
2         Algeria   NORTHERN AFRICA                        32930091   
3  American Samoa   OCEANIA                                   57794   
4         Andorra   WESTERN EUROPE                            71201   

   Area (sq. mi.) Coastline (coast/area ratio)  GDP ($ per capita)  
0          647500                         0,00               700.0  
1           28748                         1,26              4500.0  
2         2381740                         0,04              6000.0  
3             199                        58,29              8000.0  
4             468                         0,00             19000.0

### 3. Đổi tên các columns
Đổi tên các columns sau: **'Area (sq. mi.)' : 'Area' , 'Coastline (coast/area ratio)' : 'Coastline', 'GDP ($ per capita)' : 'GDP'**

In [4]:
countries_df = countries_df.rename(columns={'Area (sq. mi.)' : 'Area' , 'Coastline (coast/area ratio)' : 'Coastline', 'GDP ($ per capita)' : 'GDP'})

### 4. Xuất ra 3 bản ghi đầu tiên

In [5]:
countries_df.head(3)

Country                               Region  Population     Area  \
0  Afghanistan         ASIA (EX. NEAR EAST)             31056997   647500   
1      Albania   EASTERN EUROPE                          3581655    28748   
2      Algeria   NORTHERN AFRICA                        32930091  2381740   

  Coastline     GDP  
0      0,00   700.0  
1      1,26  4500.0  
2      0,04  6000.0

### 5. In ra các thông tin
* Chiều dài (len)
* Thông tin DataFrame (info)
* Kích thước (shape)

In [6]:
len(countries_df)

227

In [7]:
countries_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 227 entries, 0 to 226
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Country     227 non-null    object 
 1   Region      227 non-null    object 
 2   Population  227 non-null    int64  
 3   Area        227 non-null    int64  
 4   Coastline   227 non-null    object 
 5   GDP         226 non-null    float64
dtypes: float64(1), int64(2), object(3)
memory usage: 10.8+ KB


In [8]:
countries_df.shape

(227, 6)

### 6. Từ thông tin *info* của dataframe, xác định những columns nào có missing value (*null*) (chỉ cần ghi ra) và giải thích vì sao biết columns đó.
Theo thông tin *info* của dataframe ta thấy rằng DataFrame có tất cả **227** dòng, nhưng column **GDP** chỉ có **226** dòng. Điều này chứng tỏ rằng column **GDP** có missing value.

### 7. Dùng hàm *dropna()* để loại bỏ giá trị *null* đó.

In [9]:
countries_df.dropna()

Country                               Region  Population  \
0          Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1              Albania   EASTERN EUROPE                          3581655   
2              Algeria   NORTHERN AFRICA                        32930091   
3       American Samoa   OCEANIA                                   57794   
4              Andorra   WESTERN EUROPE                            71201   
..                  ...                                  ...         ...   
221  Wallis and Futuna   OCEANIA                                   16025   
222          West Bank   NEAR EAST                               2460492   
224              Yemen   NEAR EAST                              21456188   
225             Zambia   SUB-SAHARAN AFRICA                     11502010   
226           Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

        Area Coastline      GDP  
0     647500      0,00    700.0  
1      28748      1,26   4500.0  
2    2381740      0,04   6000.0  
3        199     58,29   8000.0  
4        468      0,00  19000.0  
..       ...       ...      ...  
221      274     47,08   3700.0  
222     5860      0,00    800.0  
224   527970      0,36    800.0  
225   752614      0,00    800.0  
226   390580      0,00   1900.0  

[226 rows x 6 columns]

### 8. Từ thông tin của *(5)* hãy cho biết có bao nhiêu loại *dtypes* và dùng hàm *describe* với từng loại *dtypes* đó.
Theo thông tin của **(5)** ta thấy rằng có **3 loại dtypes**:
* (1) int64
* (2) float64
* (3) object

In [10]:
countries_df.describe(include = 'int64')

Population          Area
count  2.270000e+02  2.270000e+02
mean   2.874028e+07  5.982270e+05
std    1.178913e+08  1.790282e+06
min    7.026000e+03  2.000000e+00
25%    4.376240e+05  4.647500e+03
50%    4.786994e+06  8.660000e+04
75%    1.749777e+07  4.418110e+05
max    1.313974e+09  1.707520e+07

In [11]:
countries_df.describe(include = 'float64')

GDP
count    226.000000
mean    9689.823009
std    10049.138513
min      500.000000
25%     1900.000000
50%     5550.000000
75%    15700.000000
max    55100.000000

In [12]:
countries_df.describe(include = 'object')

Country                               Region Coastline
count        227                                  227       227
unique       227                                   11       151
top     Comoros   SUB-SAHARAN AFRICA                       0,00
freq           1                                   51        44

### 9. Xuất thông tin của *'Vietnam '* nằm trong columns *'Country'*
Chú ý **'Vietnam '** có thêm dấu cách phía sau.

In [13]:
countries_vietnam_df = countries_df.loc[countries_df['Country'] == 'Vietnam ']
countries_vietnam_df

Country                         Region  Population    Area Coastline  \
219  Vietnam   ASIA (EX. NEAR EAST)             84402966  329560      1,05   

        GDP  
219  2500.0

### 10. In ra thông tin của 10 nước có *Area* lớn nhất

In [14]:
countries_top_area_df = countries_df.sort_values('Area', ascending=False).head(10)
countries_top_area_df

Country                               Region  Population  \
169         Russia                  C.W. OF IND. STATES    142893540   
36          Canada   NORTHERN AMERICA                       33098932   
214  United States   NORTHERN AMERICA                      298444215   
42           China         ASIA (EX. NEAR EAST)           1313973713   
27          Brazil               LATIN AMER. & CARIB       188078227   
11       Australia   OCEANIA                                20264082   
94           India         ASIA (EX. NEAR EAST)           1095351995   
8        Argentina               LATIN AMER. & CARIB        39921833   
106     Kazakhstan                  C.W. OF IND. STATES     15233244   
192          Sudan   SUB-SAHARAN AFRICA                     41236378   

         Area Coastline      GDP  
169  17075200      0,22   8900.0  
36    9984670      2,02  29800.0  
214   9631420      0,21  37800.0  
42    9596960      0,15   5000.0  
27    8511965      0,09   7600.0  
11    7686850      0,34  29000.0  
94    3287590      0,21   2900.0  
8     2766890      0,18  11200.0  
106   2717300      0,00   6300.0  
192   2505810      0,03   1900.0

### 11. In ra thông tin của 10 nước có *GDP* thấp nhất

In [15]:
countries_top_min_GDP_df = countries_df.sort_values('GDP').head(10)
countries_top_min_GDP_df

Country                               Region  Population  \
58             East Timor         ASIA (EX. NEAR EAST)              1062777   
188               Somalia   SUB-SAHARAN AFRICA                      8863338   
183          Sierra Leone   SUB-SAHARAN AFRICA                      6005250   
33                Burundi   SUB-SAHARAN AFRICA                      8090068   
200              Tanzania   SUB-SAHARAN AFRICA                     37445392   
74             Gaza Strip   NEAR EAST                               1428757   
125                Malawi   SUB-SAHARAN AFRICA                     13013926   
0             Afghanistan         ASIA (EX. NEAR EAST)             31056997   
46   Congo, Repub. of the   SUB-SAHARAN AFRICA                      3702314   
63                Eritrea   SUB-SAHARAN AFRICA                      4786994   

       Area Coastline    GDP  
58    15007      4,70  500.0  
188  637657      0,47  500.0  
183   71740      0,56  500.0  
33    27830      0,00  600.0  
200  945087      0,15  600.0  
74      360     11,11  600.0  
125  118480      0,00  600.0  
0    647500      0,00  700.0  
46   342000      0,05  700.0  
63   121320      1,84  700.0

### 12. Xuất ra các bản ghi có column *Region = ASIA (EX. NEAR EAST)*

In [16]:
countries_near_east_asia_region = countries_df.loc[countries_df['Region']=='ASIA (EX. NEAR EAST)         ']
countries_near_east_asia_region

Country                         Region  Population     Area  \
0     Afghanistan   ASIA (EX. NEAR EAST)             31056997   647500   
16     Bangladesh   ASIA (EX. NEAR EAST)            147365352   144000   
23         Bhutan   ASIA (EX. NEAR EAST)              2279723    47000   
29         Brunei   ASIA (EX. NEAR EAST)               379444     5770   
32          Burma   ASIA (EX. NEAR EAST)             47382633   678500   
34       Cambodia   ASIA (EX. NEAR EAST)             13881427   181040   
42          China   ASIA (EX. NEAR EAST)           1313973713  9596960   
58     East Timor   ASIA (EX. NEAR EAST)              1062777    15007   
91      Hong Kong   ASIA (EX. NEAR EAST)              6940432     1092   
94          India   ASIA (EX. NEAR EAST)           1095351995  3287590   
95      Indonesia   ASIA (EX. NEAR EAST)            245452739  1919440   
96           Iran   ASIA (EX. NEAR EAST)             68688433  1648000   
103         Japan   ASIA (EX. NEAR EAST)            127463611   377835   
109  Korea, North   ASIA (EX. NEAR EAST)             23113019   120540   
110  Korea, South   ASIA (EX. NEAR EAST)             48846823    98480   
113          Laos   ASIA (EX. NEAR EAST)              6368481   236800   
122         Macau   ASIA (EX. NEAR EAST)               453125       28   
126      Malaysia   ASIA (EX. NEAR EAST)             24385858   329750   
127      Maldives   ASIA (EX. NEAR EAST)               359008      300   
139      Mongolia   ASIA (EX. NEAR EAST)              2832224  1564116   
145         Nepal   ASIA (EX. NEAR EAST)             28287147   147181   
156      Pakistan   ASIA (EX. NEAR EAST)            165803560   803940   
162   Philippines   ASIA (EX. NEAR EAST)             89468677   300000   
184     Singapore   ASIA (EX. NEAR EAST)              4492150      693   
191     Sri Lanka   ASIA (EX. NEAR EAST)             20222240    65610   
198        Taiwan   ASIA (EX. NEAR EAST)             23036087    35980   
201      Thailand   ASIA (EX. NEAR EAST)             64631595   514000   
219       Vietnam   ASIA (EX. NEAR EAST)             84402966   329560   

    Coastline      GDP  
0        0,00    700.0  
16       0,40   1900.0  
23       0,00   1300.0  
29       2,79  18600.0  
32       0,28   1800.0  
34       0,24   1900.0  
42       0,15   5000.0  
58       4,70    500.0  
91      67,12  28800.0  
94       0,21   2900.0  
95       2,85   3200.0  
96       0,15   7000.0  
103      7,87  28200.0  
109      2,07   1300.0  
110      2,45  17800.0  
113      0,00   1700.0  
122    146,43  19400.0  
126      1,42   9000.0  
127    214,67   3900.0  
139      0,00   1800.0  
145      0,00   1400.0  
156      0,13   2100.0  
162     12,10   4600.0  
184     27,85  23700.0  
191      2,04   3700.0  
198      4,35  23400.0  
201      0,63   7400.0  
219      1,05   2500.0

### 13. Tạo một column mới có tên *'Pop. Density'* 
'Pop. Density' = 'Population' / 'Area'

In [17]:
countries_df['Pop. Density'] = countries_df['Population']/countries_df['Area']
countries_df

Country                               Region  Population  \
0       Afghanistan         ASIA (EX. NEAR EAST)             31056997   
1           Albania   EASTERN EUROPE                          3581655   
2           Algeria   NORTHERN AFRICA                        32930091   
3    American Samoa   OCEANIA                                   57794   
4           Andorra   WESTERN EUROPE                            71201   
..               ...                                  ...         ...   
222       West Bank   NEAR EAST                               2460492   
223  Western Sahara   NORTHERN AFRICA                          273008   
224           Yemen   NEAR EAST                              21456188   
225          Zambia   SUB-SAHARAN AFRICA                     11502010   
226        Zimbabwe   SUB-SAHARAN AFRICA                     12236805   

        Area Coastline      GDP  Pop. Density  
0     647500      0,00    700.0     47.964474  
1      28748      1,26   4500.0    124.587971  
2    2381740      0,04   6000.0     13.826065  
3        199     58,29   8000.0    290.422111  
4        468      0,00  19000.0    152.138889  
..       ...       ...      ...           ...  
222     5860      0,00    800.0    419.879181  
223   266000      0,42      NaN      1.026346  
224   527970      0,36    800.0     40.639029  
225   752614      0,00    800.0     15.282748  
226   390580      0,00   1900.0     31.329830  

[227 rows x 7 columns]

### 14. Thống kê số lần xuất hiện của các giá trị trong column *Region*
Dùng hàm **value_counts()**

In [18]:
countries_df.Region.value_counts()

SUB-SAHARAN AFRICA                     51
LATIN AMER. & CARIB                    45
WESTERN EUROPE                         28
ASIA (EX. NEAR EAST)                   28
OCEANIA                                21
NEAR EAST                              16
C.W. OF IND. STATES                    12
EASTERN EUROPE                         12
NORTHERN AFRICA                         6
NORTHERN AMERICA                        5
BALTICS                                 3
Name: Region, dtype: int64

### 15. Tên dư 1 dấu cách. Yêu cầu xoá dấu cách trong tên của columns Country
Dùng **apply** kết hợp **replace**

In [19]:
countries_df['Country'] = countries_df['Country'].apply(lambda x: x.replace(" ", ""))
countries_df.head()

Country                               Region  Population     Area  \
0    Afghanistan        ASIA (EX. NEAR EAST)             31056997   647500   
1        Albania  EASTERN EUROPE                          3581655    28748   
2        Algeria  NORTHERN AFRICA                        32930091  2381740   
3  AmericanSamoa  OCEANIA                                   57794      199   
4        Andorra  WESTERN EUROPE                            71201      468   

  Coastline      GDP  Pop. Density  
0      0,00    700.0     47.964474  
1      1,26   4500.0    124.587971  
2      0,04   6000.0     13.826065  
3     58,29   8000.0    290.422111  
4      0,00  19000.0    152.138889

### 16. Lưu dataframe với tên *'countries_of_the_world.csv'*

In [20]:
countries_df.to_csv('countries_of_the_world.csv')

### Bonus
Ta thấy dtype của columns Coastline là object, nhưng khi nhìn vào giá trị thì ta sẽ nghĩ tới là numerical.
Yêu cầu: 
* Chuyển đổi dtype của cột **Coastline** sang **numerical(float)**, chú ý dấu phẩy ','.
* Sau đó, cho biết các nước (**Country**) không tiếp giáp với biển (**coastline = 0**).

In [21]:
print(countries_df['Coastline'].dtypes)
countries_df['Coastline'].head()

object


0     0,00
1     1,26
2     0,04
3    58,29
4     0,00
Name: Coastline, dtype: object

In [22]:
countries_df['Coastline'] = countries_df['Coastline'].apply(lambda x: x.replace(",", "."))
countries_df['Coastline'].head()

0     0.00
1     1.26
2     0.04
3    58.29
4     0.00
Name: Coastline, dtype: object

In [23]:
countries_df['Coastline'] = countries_df['Coastline'].apply(pd.to_numeric, errors='coerce')
countries_df['Coastline'].head()

0     0.00
1     1.26
2     0.04
3    58.29
4     0.00
Name: Coastline, dtype: float64

In [24]:
countries_df['Country'].loc[countries_df['Coastline']==0]

0             Afghanistan
4                 Andorra
9                 Armenia
12                Austria
13             Azerbaijan
18                Belarus
23                 Bhutan
24                Bolivia
26               Botswana
31            BurkinaFaso
33                Burundi
39     CentralAfricanRep.
40                   Chad
45         Congo,Dem.Rep.
53          CzechRepublic
65               Ethiopia
92                Hungary
106            Kazakhstan
112            Kyrgyzstan
113                  Laos
116               Lesotho
119         Liechtenstein
121            Luxembourg
123             Macedonia
125                Malawi
128                  Mali
137               Moldova
139              Mongolia
145                 Nepal
151                 Niger
160              Paraguay
170                Rwanda
177             SanMarino
181                Serbia
185              Slovakia
194             Swaziland
196           Switzerland
199            Tajikistan
207         